src/scvi/module/_utils.py

```diff
@@ -1,45 +1,31 @@
 import torch
 
 
-def iterate(obj, func):
-    """Iterates over an object and applies a function to each element."""
-    t = type(obj)
-    if t is list or t is tuple:
-        return t([iterate(o, func) for o in obj])
-    else:
-        return func(obj) if obj is not None else None
-
-
-def broadcast_labels(y, *o, n_broadcast=-1):
+def broadcast_labels(o, n_broadcast=-1):
     """Utility for the semi-supervised setting.
 
     If y is defined(labelled batch) then one-hot encode the labels (no broadcasting needed)
     If y is undefined (unlabelled batch) then generate all possible labels (and broadcast other
     arguments if not None)
     """
-    if not len(o):
-        raise ValueError("Broadcast must have at least one reference argument")
-    if y is None:
-        ys = enumerate_discrete(o[0], n_broadcast)
-        new_o = iterate(
-            o,
-            lambda x: x.repeat(n_broadcast, 1) if len(x.size()) == 2 else x.repeat(n_broadcast),
-        )
-    else:
-        ys = torch.nn.functional.one_hot(y.squeeze(-1), n_broadcast)
-        new_o = o
-    return (ys,) + new_o
-
-
-def enumerate_discrete(x, y_dim):
-    """Enumerate discrete variables."""
 
     def batch(batch_size, label):
-        labels = torch.ones(batch_size, 1, device=x.device, dtype=torch.long) * label
-        return torch.nn.functional.one_hot(labels.squeeze(-1), y_dim)
-
-    batch_size = x.size(0)
-    return torch.cat([batch(batch_size, i) for i in range(y_dim)])
+        labels = torch.ones(batch_size, 1, device=o.device, dtype=torch.long) * label
+        # return size (batch_size, n_broadcast)
+        return torch.nn.functional.one_hot(labels.squeeze(-1), n_broadcast)
+
+    if o.ndim == 2:
+        batch_size = o.size(0)
+        ys = torch.cat([batch(batch_size, i) for i in range(n_broadcast)])
+        new_o = o.repeat(n_broadcast, 1)
+    elif o.ndim == 3:
+        n_samples = o.size(0)
+        batch_size = o.size(1)
```

Comment:
o.size(-2) should work for both.

Reply:
not so sure what you meant. o.size(-2) gave me 128 which is like o.size(1) but

Reply:
You call here batch_size = o.size(1) and for ndim=2 batch_size = o.size(0). It's nicer to call o.size(-2) and do it outside of the if statement.

Reply:
got it.